<a href="https://colab.research.google.com/github/ianzin30/IF1014-Mining/blob/main/Nomao.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# CRISP_DM PHASE 1
Entendimento

https://docs.google.com/document/d/1__5aHtUM-75B60eRaHWH0kN8GaWUBLRkCMg_DDWnB0E/edit?usp=sharing



# CRISP_DM PHASE 2
Compreensão

## Inicializando dados


In [ ]:
# Executando os imports necessários para rodar o projeto

import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
# Adquirindo features do Nomao
file_path_features = '/content/drive/MyDrive/nomao_dataset/Nomao.features'

with open(file_path_features, 'r') as file:
    content_features = file.read()

# Adquirindo data do Nomao
file_path_data = '/content/drive/MyDrive/nomao_dataset/Nomao.data'
with open(file_path_data, 'r') as file:
    content_data = file.read()

# Colocando features e data para criar o dataFrame
features = content_features.split('\n')
data = list(map(lambda x: x.split(','), content_data.split('\n')))

In [ ]:
# Inicializando DataFrame do Nomao
nomao_df = pd.DataFrame(data, columns=features)
nomao_df.head()

id clean_name_intersect_min clean_name_intersect_max  \
0  0#1                        1                        1   
1  0#2                        1                     0.75   
2  0#3                        1                        1   
3  4#5                        1                     0.75   
4  6#7                        0                        0   

  clean_name_levenshtein_sim clean_name_trigram_sim  \
0                          1                      1   
1                   0.857143               0.857143   
2                          1                      1   
3                   0.857143               0.857143   
4                       0.25                      0   

  clean_name_levenshtein_term clean_name_trigram_term clean_name_including  \
0                           1                       1                    s   
1                    0.894737                0.947368                    s   
2                           1                       1                    s   
3                    0.842105                0.833333                    s   
4                        0.25                       0                    n   

  clean_name_equality city_intersect_min  ... coordinates_long_levenshtein  \
0                   s                  ?  ...                         0.75   
1                   n                  ?  ...                          0.5   
2                   s                  ?  ...               0.666666666667   
3                   n                  1  ...                          0.5   
4                   n                  ?  ...                            ?   

  coordinates_long_trigram coordinates_long_equality coordinates_lat_diff  \
0                      0.5                         n       0.999953002209   
1           0.307692319155                         n       0.999982921892   
2           0.461538463831                         n       0.999992600055   
3           0.285714298487                         n       0.998604948886   
4                        ?                         m                    ?   

  coordinates_lat_levenshtein coordinates_lat_trigram  \
0              0.777777777778          0.461538463831   
1              0.538461538462                  0.4375   
2                         0.8          0.666666686535   
3              0.555555555556          0.384615391493   
4                           ?                       ?   

  coordinates_lat_equality geocode_coordinates_diff coordinates_diff label  
0                        n                      1.0   0.999999996702    +1  
1                        n                      1.0   0.999999962459    +1  
2                        n                      1.0   0.999999999653    +1  
3                        n           0.999999947284   0.999994124065    +1  
4                        m                      1.0                ?    +1  

[5 rows x 120 columns]

## Entendendo melhor os dados do Nomao



Segundo o arquivo de informação do desafio da Nomao, é informado que:

First 29,104 instances have been labeled with "human prior".

Next 917 instances have been labeled using the active learning method called "marg".

Next 964 instances refer to the active method called "wmarg".

Next 995 instances refer to the active method called "wmarg5".

Next 1,985 instances refer to the active method called "rand" (random selection).

Next 163 instances refer to the active method called "baseline".

Next 167 instances refer to the active method called "nomao".

And last 170 instances refer to the active method called "tsun".


In [ ]:
nomao_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34466 entries, 0 to 34465
Columns: 120 entries, id to label
dtypes: object(120)
memory usage: 31.6+ MB


Podemos ver que o método "info", nos diz que existem 34466 linhas e 120 features, um arquivo bem grande e que condiz com o que foi informado pelo desafio.

Podemos então começar com o pré processamento de dados, verificando duplicatas, linhas com informações faltantes, inconcistência.

Vale ressaltar que no arquivo .names, é informado como é catégorizado dados incompletos, "?"


In [ ]:
nomao_df.duplicated().sum()

0

In [ ]:
nomao_df.dtypes

id                            object
clean_name_intersect_min      object
clean_name_intersect_max      object
clean_name_levenshtein_sim    object
clean_name_trigram_sim        object
                               ...  
coordinates_lat_trigram       object
coordinates_lat_equality      object
geocode_coordinates_diff      object
coordinates_diff              object
label                         object
Length: 120, dtype: object

In [ ]:
# Não é necessário rodar, mas é interessante para ver a quantidade de NaN em cada coluna do Dataset
pd.set_option('display.max_rows', None)

In [ ]:
nomao_df.replace("?", pd.NA, inplace=True) # Irei transformar todos os valores "?" com o padrão de Missing value de Pandas.
x = nomao_df.isna().sum()
x.sort_values(ascending=False)
# Tem muita coluna com MUITOS valores em NaN. Por mim tira


fax_diff                                     33570
fax_levenshtein                              33570
fax_trigram                                  33570
website_intersect_max                        32289
website_intersect_min                        32289
website_trigram_sim                          32289
website_levenshtein_term                     32289
website_trigram_term                         32289
website_levenshtein_sim                      32289
countryname_levenshtein_sim                  28798
countryname_intersect_min                    28798
countryname_intersect_max                    28798
countryname_levenshtein_term                 28798
countryname_trigram_sim                      28798
countryname_trigram_term                     28798
zip_levenshtein_sim                          25111
zip_trigram_term                             25111
zip_levenshtein_term                         25111
zip_trigram_sim                              25111
zip_intersect_min                            25111
zip_intersect_max                            25111
street_intersect_min                         24446
street_intersect_max                         24446
street_levenshtein_sim                       24446
street_trigram_sim                           24446
street_levenshtein_term                      24446
street_trigram_term                          24446
city_intersect_max                           22296
city_levenshtein_sim                         22296
city_intersect_min                           22296
city_trigram_term                            22296
city_trigram_sim                             22296
city_levenshtein_term                        22296
phone_levenshtein                            20054
phone_trigram                                20054
phone_diff                                   20054
coordinates_lat_trigram                      18966
coordinates_lat_diff                         18966
coordinates_diff                             18966
coordinates_long_diff                        18966
coordinates_long_levenshtein                 18966
coordinates_long_trigram                     18966
coordinates_lat_levenshtein                  18966
street_number_trigram                        11986
street_number_diff                           11986
street_number_levenshtein                    11986
geocoderpostalcodenumber_intersect_min        4570
geocoderpostalcodenumber_intersect_max        4570
geocoderpostalcodenumber_levenshtein_sim      4570
geocoderpostalcodenumber_trigram_sim          4570
geocoderpostalcodenumber_levenshtein_term     4570
geocoderpostalcodenumber_trigram_term         4570
geocodercountrynamecode_trigram_sim           1354
geocodercountrynamecode_levenshtein_sim       1354
geocodercountrynamecode_trigram_term          1354
geocodercountrynamecode_intersect_min         1354
geocodercountrynamecode_levenshtein_term      1354
geocodercountrynamecode_intersect_max         1354
geocoderlocalityname_trigram_term              540
geocoderlocalityname_levenshtein_term          540
geocoderlocalityname_trigram_sim               540
geocoderlocalityname_levenshtein_sim           540
geocoderlocalityname_intersect_min             540
geocoderlocalityname_intersect_max             540
geocoderinputaddress_trigram_sim                56
geocoderinputaddress_intersect_min              56
geocoderinputaddress_trigram_term               56
geocoderinputaddress_intersect_max              56
geocoderinputaddress_levenshtein_sim            56
geocoderinputaddress_levenshtein_term           56
geocoderoutputaddress_intersect_max             10
geocoderoutputaddress_intersect_min             10
geocoderoutputaddress_trigram_term              10
geocoderoutputaddress_levenshtein_term          10
geocoderoutputaddress_trigram_sim               10
geocoderoutputaddress_levenshtein_sim           10
geocode_coordinates_lat_trigram                  7
geocode_coordinates_long_levenshtein             7
geocode_coordinates_la

In [ ]:
nomao_df['fax_diff'].head()

0    ?
1    ?
2    ?
3    ?
4    ?
Name: fax_diff, dtype: object

In [ ]:
nomao_df['clean_name_intersect_']

Index(['id', 'clean_name_intersect_min', 'clean_name_intersect_max',
       'clean_name_levenshtein_sim', 'clean_name_trigram_sim',
       'clean_name_levenshtein_term', 'clean_name_trigram_term',
       'clean_name_including', 'clean_name_equality', 'city_intersect_min',
       ...
       'coordinates_long_levenshtein', 'coordinates_long_trigram',
       'coordinates_long_equality', 'coordinates_lat_diff',
       'coordinates_lat_levenshtein', 'coordinates_lat_trigram',
       'coordinates_lat_equality', 'geocode_coordinates_diff',
       'coordinates_diff', 'label'],
      dtype='object', length=120)